### PR TITLE
EFI: added efi MavLink class

### PR DESCRIPTION
**Pull request by Max-001**
I added a MavLink class to the EFI library to store EFI-data that is received over MavLink. There are EFI systems that don't use the Cube as their "computer", those systems have their own "computer" and forward the EFI-data over MavLink. It can be usefull to store this data (using the EFI library). For example: If the EFI-data is stored, it can accessed using a LUA script. The accessing of EFI-data using a LUA script is already programmed in to the main branch.

The new class called AP_EFI_MAV has a method called handle_EFI_message, that's the method that handles the MavLink data and stores it. This method gets called in GCS_Common.cpp at the moment that a MAVLINK_MSG_ID_EFI_STATUS message appears (this is done using the switch-case in the method GCS_MAVLINK::handle_common_message). To make this all work, the EFI_TYPE parameter has to be set to 9 (meaning MAV(link)).

As an example I have writen this LUA script:
```
function update()

  local EFI_state = efi:get_state()

  gcs:send_text(6, EFI_state:intake_manifold_temperature())
  gcs:send_text(6, EFI_state:throttle_out())
  gcs:send_text(6, EFI_state:cylinder_status():cylinder_head_temperature())

  -- run at 1hz
  return update, 10000
end

return update()
```

The result of this LUA script (using the code in this pull-request) is shown below:
image

You can see the incomming MavLink data of the EFI system on the MavLink inspector and you can see the output of the LUA script on the left side of the screenshot. Please notice that the shown data on the MavLink inspector is in Celsius while the data shown by the script is in Kelvin.

Please refer to this earlier pull request: https://github.com/ArduPilot/ardupilot/pull/24113

### CODE REVIEW

File: libraries/AP_EFI/AP_EFI_config.h [hunk 1]
```diff
@@ -34,3 +34,7 @@
 #ifndef AP_EFI_SERIAL_LUTAN_ENABLED
 #define AP_EFI_SERIAL_LUTAN_ENABLED AP_EFI_BACKEND_DEFAULT_ENABLED
 #endif
+
+#ifndef AP_EFI_MAV_ENABLED
```
Comment (peterbarker) [inline libraries/AP_EFI/AP_EFI_config.h:38]: Alphabetical order in here, by the looks of it.

File: libraries/AP_EFI/AP_EFI.cpp [hunk 2]
```diff
@@ -117,6 +118,11 @@ void AP_EFI::init(void)
         backend = new AP_EFI_Scripting(*this);
 #endif
         break;
+    case Type::MAV:
+#if AP_EFI_MAV_ENABLED
```
Comment (peterbarker) [inline libraries/AP_EFI/AP_EFI.cpp:122]: #if around the `case`, please

**Review by peterbarker — CHANGES_REQUESTED** [review]
Please roll the other PR into this one.

File: libraries/AP_EFI/AP_EFI_MAV.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+//Decode MavLink message
+void AP_EFI_MAV::handle_EFI_message(const mavlink_message_t &msg)
+{
+    mavlink_efi_status_t state;
+    mavlink_msg_efi_status_decode(&msg, &state);
+
+    internal_state.ecu_index = state.ecu_index;
+    //internal_state.??? = state.rpm;
```
Comment (peterbarker) [inline libraries/AP_EFI/AP_EFI_MAV.cpp:37]: Check my PR for mappings for this.

https://github.com/ArduPilot/ardupilot/commit/8d52efbf8c44353012455d66926a8d99cfa452a0#diff-fd3b580d602da071426c9e0387b1cc0283213a5d883b00a5ae7cf89467ea4e09R15

**Review by peterbarker — APPROVED** [review]

**Comment by IamPete1** [comment]
This could be done in completely scripting with 0 flash cost now we have MAVLink support.

**Comment by Max-001** [comment]
Thank you for the approval @peterbarker !

I have done some more testing, besides the test that is described above (please see the first comment). I tested everything using a real EFI system. I compared the MavLink inspector with the outcome of the following LUA script:
```
function update()

  local EFI_state = efi:get_state()

  gcs:send_text(6, "ecu_index: " .. EFI_state:ecu_index())
  gcs:send_text(6, "estimated_consumed_fuel_volume_cm3: " .. EFI_state:estimated_consumed_fuel_volume_cm3())
  gcs:send_text(6, "fuel_consumption_rate_cm3pm: " .. EFI_state:fuel_consumption_rate_cm3pm())
  gcs:send_text(6, "engine_load_percent: " .. EFI_state:engine_load_percent())
  gcs:send_text(6, "throttle_position_percent: " .. EFI_state:throttle_position_percent())
  gcs:send_text(6, "spark_dwell_time_ms: " .. EFI_state:spark_dwell_time_ms())
  gcs:send_text(6, "atmospheric_pressure_kpa: " .. EFI_state:atmospheric_pressure_kpa())
  gcs:send_text(6, "intake_manifold_pressure_kpa: " .. EFI_state:intake_manifold_pressure_kpa())
  gcs:send_text(6, "intake_manifold_temperature: " .. EFI_state:intake_manifold_temperature())
  gcs:send_text(6, "cylinder_head_temperature: " .. EFI_state:cylinder_status():cylinder_head_temperature())
  gcs:send_text(6, "ignition_timing_deg: " .. EFI_state:cylinder_status():ignition_timing_deg())
  gcs:send_text(6, "injection_time_ms: " .. EFI_state:cylinder_status():injection_time_ms())
  gcs:send_text(6, "exhaust_gas_temperature: " .. EFI_state:cylinder_status():exhaust_gas_temperature())
  gcs:send_text(6, "throttle_out: " .. EFI_state:throttle_out())
  gcs:send_text(6, "pt_compensation: " .. EFI_state:pt_compensation())
  gcs:send_text(6, "ignition_voltage: " .. EFI_state:ignition_voltage())

  -- run at 1hz
  return update, 10000
end

return update()
```

The MavLink inspector and the outcome of the LUA script are the same, please see the screenshot below. There are a few very small differences between the MavLink inspector and the outcome of the LUA script, that is because the MavLink inspector is updated much quicker than the LUA script (so the small differences is nothing to worry about).
screenshot3

I also compared the MavLink inspector and the outcome of the LUA script for a longer period of time and the values are continuously the same. This is done with and without running motor.

**Review by rmackay9 — APPROVED** [review]

**Comment by AliSeedg** [comment]
Can you give me the code you used?

**Comment by Max-001** [comment]
> Can you give me the code you used?

Do you mean the LUA script?

```
function update()

  local EFI_state = efi:get_state()

  gcs:send_text(6, "ecu_index: " .. EFI_state:ecu_index())
  gcs:send_text(6, "estimated_consumed_fuel_volume_cm3: " .. EFI_state:estimated_consumed_fuel_volume_cm3())
  gcs:send_text(6, "fuel_consumption_rate_cm3pm: " .. EFI_state:fuel_consumption_rate_cm3pm())
  gcs:send_text(6, "engine_load_percent: " .. EFI_state:engine_load_percent())
  gcs:send_text(6, "throttle_position_percent: " .. EFI_state:throttle_position_percent())
  gcs:send_text(6, "spark_dwell_time_ms: " .. EFI_state:spark_dwell_time_ms())
  gcs:send_text(6, "atmospheric_pressure_kpa: " .. EFI_state:atmospheric_pressure_kpa())
  gcs:send_text(6, "intake_manifold_pressure_kpa: " .. EFI_state:intake_manifold_pressure_kpa())
  gcs:send_text(6, "intake_manifold_temperature: " .. EFI_state:intake_manifold_temperature())
  gcs:send_text(6, "cylinder_head_temperature: " .. EFI_state:cylinder_status():cylinder_head_temperature())
  gcs:send_text(6, "ignition_timing_deg: " .. EFI_state:cylinder_status():ignition_timing_deg())
  gcs:send_text(6, "injection_time_ms: " .. EFI_state:cylinder_status():injection_time_ms())
  gcs:send_text(6, "exhaust_gas_temperature: " .. EFI_state:cylinder_status():exhaust_gas_temperature())
  gcs:send_text(6, "throttle_out: " .. EFI_state:throttle_out())
  gcs:send_text(6, "pt_compensation: " .. EFI_state:pt_compensation())
  gcs:send_text(6, "ignition_voltage: " .. EFI_state:ignition_voltage())

  -- run at 1hz
  return update, 10000
end

return update()
```

**Comment by AliSeedg** [comment]
Thank you really
But I have a problem. I send an efi status message via Arduino and it appears to me in the mavlink inspector, but it does not appear in quick. I want a solution. Is it by changing the main system code or via lua?

**Comment by AliSeedg** [comment]
06/02/24 06:28:39 م : ll a nil value (method 'get_state')
06/02/24 06:28:39 م : Lua: /APM/scripts/MAX.lua:3: attempt to ca
I see this when I use your code

**Comment by AliSeedg** [comment]
Yes

**Comment by Max-001** [comment]
> Yes

Ok, and I assume that it still doesn’t work.
Could you please share you’re setup in more detail (e.g. hardware connection from Arduino to Cube, Arduino code, ArduPilot setup, etcetera)?

**Comment by AliSeedg** [comment]
Thank you. My problem was solved and my data appeared on the quick screen without using the Lua language, only via Arduino and Maflink. The problem was in the firmware file.

**Pull request merged**